### PR TITLE
AFHS-599 Checking eligibility override when evaluating eligibility for a new claim

### PR DESCRIPTION
**Pull request by YetAnotherMatt**

### CODE REVIEW

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementService.java [hunk 1]
```diff
@@ -80,4 +87,23 @@ public EligibilityAndEntitlementDecision evaluateClaimantForPaymentCycle(Claiman
...
+                    .addressLine1Match(VerificationOutcome.NOT_SET)
+                    .deathVerificationFlag(DeathVerificationFlag.N_A)
+                    .dobOfChildrenUnder4(Collections.emptyList())
+                    .emailAddressMatch(VerificationOutcome.NOT_SET)
+                    .mobilePhoneMatch(VerificationOutcome.NOT_SET)
+                    .postcodeMatch(VerificationOutcome.NOT_SET)
+                    .pregnantChildDOBMatch(VerificationOutcome.MATCHED)
+                    .build();
+        }
+    }
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementService.java]: I've had a bit of a tinker with this as it looks a bit clunky, how about this as an alternative?
`    private CombinedIdentityAndEligibilityResponse getCombinedIdentityAndEligibilityResponse(Claimant claimant,
                                                                                             EligibilityOutcome eligibilityOverrideOutcome) {
        if(isOverride(eligibilityOverrideOutcome)) {
            return buildOverridenResponse(eligibilityOverrideOutcome);
        }
        return return client.checkIdentityAndEligibility(claimant);
    }

    private boolean isOverride(EligibilityOutcome eligibilityOverrideOutcome) {
        return eligibilityOverrideOutcome != null;
    }

    private CombinedIdentityAndEligibilityResponse buildOverridenResponse(EligibilityOutcome eligibilityOverrideOutcome) {
        return CombinedIdentityAndEligibilityResponse.builder()
                .identityStatus(IdentityOutcome.MATCHED)
                .eligibilityStatus(eligibilityOverrideOutcome)
                .addressLine1Match(VerificationOutcome.NOT_SET)
                .deathVerificationFlag(DeathVerificationFlag.N_A)
                .dobOfChildrenUnder4(Collections.emptyList())
                .emailAddressMatch(VerificationOutcome.NOT_SET)
                .mobilePhoneMatch(VerificationOutcome.NOT_SET)
                .postcodeMatch(VerificationOutcome.NOT_SET)
                .pregnantChildDOBMatch(VerificationOutcome.MATCHED)
                .build();
    }`

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimService.java [hunk 2]
```diff
@@ -171,6 +172,7 @@ private Claim buildClaim(ClaimStatus claimStatus, ClaimRequest claimRequest, Eli
                 .webUIVersion(claimRequest.getWebUIVersion())
                 .initialIdentityAndEligibilityResponse(decision.getIdentityAndEligibilityResponse())
                 .currentIdentityAndEligibilityResponse(decision.getIdentityAndEligibilityResponse())
+                .eligibilityOverrideOutcome(claimRequest.getEligibilityOverrideOutcome())
```
Comment (mikeSmith08) [inline api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimService.java]: An open question: do we need to store this after the override has been applied, is this here just for record keeping? Feels like that might be better stored in the CRM system (and yes I do mean the CRM system).

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java [hunk 3]
```diff
@@ -41,6 +50,7 @@
 
     private static final boolean NOT_DUPLICATE = false;
     private static final boolean DUPLICATE = true;
+    private  static final EligibilityOutcome NO_ELIGIBILITY_OVERRIDE = null;
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: Double space here, surprised that got through Checkstyle.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java [hunk 4]
```diff
@@ -64,21 +74,30 @@
...
-        EligibilityAndEntitlementDecision decision = eligibilityAndEntitlementService.evaluateNewClaimant(CLAIMANT);
+        EligibilityAndEntitlementDecision decision = eligibilityAndEntitlementService.evaluateNewClaimant(CLAIMANT, eligibilityOverrideOutcome);
 
         //Then
         assertThat(decision.getEligibilityStatus()).isEqualTo(EligibilityStatus.DUPLICATE);
         assertThat(decision.getExistingClaimId()).isEqualTo(existingClaimId);
         verifyNoInteractions(client);
     }
 
+    private static Stream<EligibilityOutcome> allEligibilityOutcomeValues() {
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: Technically this isn't all eligibility outcomes as we're not testing NOT_SET, do we want to add handling for this value? If not, we should probably rename this method accordingly, something like `allValidEligibilityOutcomes`.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java [hunk 5]
```diff
@@ -103,14 +122,42 @@ void shouldReturnEligibleWhenNotDuplicateAndEligible() {
...
+        given(paymentCycleEntitlementCalculator.calculateEntitlement(any(), any(), any())).willReturn(VOUCHER_ENTITLEMENT);
+        given(claimRepository.findLiveClaimWithNino(any())).willReturn(Optional.empty());
+        given(duplicateClaimChecker.liveClaimExistsForHousehold(any(CombinedIdentityAndEligibilityResponse.class))).willReturn(NOT_DUPLICATE);
+        EligibilityAndEntitlementDecision decisionResponse = setupEligibilityAndEntitlementDecisionFactory(ELIGIBLE);
+
+        //When
+        EligibilityAndEntitlementDecision result = eligibilityAndEntitlementService.evaluateNewClaimant(CLAIMANT, EligibilityOutcome.CONFIRMED);
+
+        //Then
+        CombinedIdentityAndEligibilityResponse response = CombinedIdentityAndEligibilityResponse.builder()
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: I think this would be a good candidate to put in a TestDataFactory as `aConfirmedOverrridenResponse` or `anOverridenResponse` taking in the `EligibilityOutcome`.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimServiceTest.java [hunk 6]
```diff
@@ -42,48 +42,32 @@
...
-import static uk.gov.dhsc.htbhf.eligibility.model.EligibilityStatus.INELIGIBLE;
+import static uk.gov.dhsc.htbhf.dwp.model.VerificationOutcome.*;
+import static uk.gov.dhsc.htbhf.eligibility.model.EligibilityStatus.*;
 import static uk.gov.dhsc.htbhf.eligibility.model.testhelper.CombinedIdAndEligibilityResponseTestDataFactory.*;
 
 @ExtendWith(MockitoExtension.class)
 class ClaimServiceTest {
 
     private static final List<LocalDate> NULL_CHILDREN = null;
+    private  static final EligibilityOutcome NO_ELIGIBILITY_OVERRIDE = null;
```
Comment (mikeSmith08) [inline api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimServiceTest.java]: Double space here too.

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementService.java (hunk 1 above)
Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementService.java]: applied, thanks

File: api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimService.java (hunk 2 above)
Comment (YetAnotherMatt) [reply api/src/main/java/uk/gov/dhsc/htbhf/claimant/service/claim/ClaimService.java]: Yes, it will need to be applied in every payment cycle - so it needs to be stored with the claim.

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java (hunk 4 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: updated with all outcomes

Comment (dwybourn) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: Should we use an EnumSource and another test for null?

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java (hunk 3 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: done

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java (hunk 5 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: done

File: api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java (hunk 4 above)
Comment (YetAnotherMatt) [reply api/src/test/java/uk/gov/dhsc/htbhf/claimant/service/EligibilityAndEntitlementServiceTest.java]: as discussed, I'm avoiding that because we're only testing one thing - that all values of EligibilityOutcome are ignored when there is a duplicate nino.
